Add atoum task

jhuet opened the pull request:
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Documented?   | yes
| Fixed tickets | no

[Atoum](http://atoum.org/) is unit test framework that could totally be run in a grumphp task. This is an implementation of its usage containing specs and doc.

veewee commented on src/GrumPHP/Task/Atoum.php:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * @return OptionsResolver
+     */
+    public function getConfigurableOptions()
```
Shouldn't the filters be arrays instead of strings? In that case, you could set an empty array as default value instead.

jhuet replied:
See my comments below.

veewee commented on src/GrumPHP/Task/Atoum.php:
```diff
@@ -0,0 +1,103 @@
+        $config = $this->getConfiguration();
+
+        $arguments = $this->processBuilder->createArgumentsForCommand('atoum');
+        if ($config['config_file']) {
```
For this kind of if structures, there is an `addOptionalArgument` method on the `ProcessArgumentsCollection` class.

jhuet replied:
The thing is that the actual parameter syntax for atoum is this one : `atoum --bootstrap-file tests/units/bootstrap.php` but if i use `addOptionalArgument('--bootstrap-file %s', $config['bootstrap_file'])`, the `Process` will escape the whole argument making it be like : `'atoum' '--bootstrap-file tests/units/bootstrap.php'` instead of `'atoum' '--bootstrap-file' 'tests/units/bootstrap.php'`. I'm not sure how to do this without using `addArgumentArrayWithSeparatedValue()` which is why i kinda tricked around using it.

veewee commented on src/GrumPHP/Task/Atoum.php:
```diff
@@ -0,0 +1,103 @@
+        if ($config['methods']) {
+            $arguments->addArgumentArrayWithSeparatedValue('--methods', array($config['methods']));
+        }
+        if ($config['tags']) {
```
If the tags section is an array in the configuration, you can reomove these ifs and directly run the `addArgumentArrayWithSeparatedValue()` method.

jhuet replied:
I have a similar problem here. For multiple arguments, the syntax is : `atoum --directories dir1 dir2`. If i use `addOptionalArgument` or like now `addArgumentArrayWithSeparatedValue` by "faking" a string, all directories are escaped in 1 parameter : `'atoum' '--directories' 'dir1 dir2'`. If i use `addArgumentArrayWithSeparatedValue` with an actual array, escaping is ok but the parameter name is duplicated : `'atoum' '--directories' 'dir1' '--directories' 'dir2'`. If you have any advice on how to do this, it'd be very much welcome :)

veewee commented:
Hi @jhuet,

I am not familiar with `atoum` so I am not sure how the cli configuration works exactly. 
I've added some small remarks to the code. 

Thanks for the new task!

veewee commented:
Hi @jhuet,

That way of adding arguments is pretty common.
Maybe you can just add 2 new methods to the `ProcessArgumentsCollection`?
I was thinking about something like:

`addOptionalSeparatedArgument()`

This method will look like `addOptionalArgument` but will add 2 arguments instead of a sprintf if the value is available.


`addSeparatedArgumentArray()`

This method will look like `addArgumentArray` but will set the key once and multiple value arguments. Note that when the array is empty, the key of the argument should not be set.

This way you can make the task more readable and improve the types of the configurable options.

jhuet commented:
Alright @veewee, i'll find some time to add these methods. For some reason i didn't realize this class was part of grumphp, i thought it was symfony's :)

veewee commented:
@jhuet Are you planning to finish this PR any time soon? Thanks!

jhuet commented:
@veewee, sorry i've been busy and i'm now on vacations. I'll try to finish this during the week-end of the 17th of september.

veewee commented:
Have a great vacation! ;)

jhuet commented:
Hey @veewee, i finally pushed an updated version of this task using the 2 new methods `addOptionalSeparatedArgument()` and `addSeparatedArgumentArray()` that i added in `ProcessArgumentsCollection`.

I also rebased all of this from the master branch.

veewee commented on src/GrumPHP/Collection/ProcessArgumentsCollection.php:
```diff
@@ -65,6 +65,18 @@ public function addOptionalCommaSeparatedArgument($argument, array $values, $del
      * @param string $argument
+     * @param string $value
+     */
+    public function addOptionalSeparatedArgument($argument, $value = null)
```
This method was already implemented in the `addOptionalArgumentWithSeparatedValue` method:
https://github.com/jhuet/grumphp/blob/8ff4bb0ff833d00ad0ea010cc164961832173df7/src/GrumPHP/Collection/ProcessArgumentsCollection.php#L42

Can you remove the `addOptionalSeparatedArgument` method and use the one above instead??

veewee commented on src/GrumPHP/Collection/ProcessArgumentsCollection.php:
```diff
@@ -90,6 +102,22 @@ public function addArgumentArrayWithSeparatedValue($argument, array $values)
      * @param string $argument
+     * @param array  $values
+     */
+    public function addSeparatedArgumentArray($argument, array $values)
```
Can you add phpspec tests for this method?

veewee commented on src/GrumPHP/Task/Atoum.php:
```diff
@@ -0,0 +1,89 @@
+     */
+    public function run(ContextInterface $context)
+    {
+        $files = $context->getFiles()->name('*.php');
```
I haven't worked with atoum myself. Is it possible to add the `$files` parameter to the process or will it always run all of its tests?

jhuet replied:
Yes, Atoum allows to run tests for specific files but `$files` would have to be the test files. AFAIK, if a file that is unit tested is given to Atoum command line, it won't know which tests to execute. So i'm not sure it would be very usefull to use.

veewee replied:
That's what I thought. Great! :)

veewee commented on doc/tasks/atoum.md:
```diff
@@ -0,0 +1,81 @@
+
+**directories**
+
+*Default: array()*
```
In other documentation files we use `[]` instead of `array()`. Can you update the documentation so it looks the same like the other tasks?

veewee commented:
Thanks for your work!
